### PR TITLE
fix(ssa): synchronisation for functions

### PR DESCRIPTION
**Pull request by guipublic**
# Description

## Summary of changes

This PR fixes the synchronisation of ELSE/THEN branches. As a result, we now need to properly optimise result instructions coming from the branches.

## Test additions / changes

Test case added in 9_conditional

# Checklist

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` with default settings.
- [ ] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this PR to the issue(s) that it resolves.
- [X] I have reviewed the changes on GitHub, line by line.
- [X] I have ensured all changes are covered in the description.
- [ ] This PR requires documentation updates when merged.

### CODE REVIEW

File: crates/noirc_evaluator/src/ssa/optimizations.rs
Comment (kevaundray) [inline crates/noirc_evaluator/src/ssa/optimizations.rs:243]: why is this a debug_assert instead of a regular assert?

Comment (guipublic) [reply crates/noirc_evaluator/src/ssa/optimizations.rs:243]: For no real reason, it could be a regular one as it is not supposed to happen. A debug_assert! has no impact so it is harmless and I supposed I added it during debugging. I will be happy to change it to a regular one if you prefer.

Comment (kevaundray) [reply crates/noirc_evaluator/src/ssa/optimizations.rs:249]: A debug_assert gets removed on Release builds, so I was a bit confused :) I'd suggest we use assert! all over the codebase for things that shouldn't happen so they don't get optimised out, other than that happy to merge

Comment (guipublic) [reply crates/noirc_evaluator/src/ssa/optimizations.rs]: Done

File: crates/noirc_evaluator/src/ssa/optimizations.rs
Comment (jfecher) [inline crates/noirc_evaluator/src/ssa/optimizations.rs:236]: If we're asserting `!=`, we should use `assert_ne` instead since this will print out both operands if they differ

File: crates/noirc_evaluator/src/ssa/context.rs
Comment (jfecher) [inline crates/noirc_evaluator/src/ssa/context.rs:705]: Can you elaborate on why we need to do a full_cse twice now?

File: crates/noirc_evaluator/src/ssa/conditional.rs
Comment (jfecher) [inline crates/noirc_evaluator/src/ssa/conditional.rs:379]: Can you elaborate on why we need to cse twice here as well? Why is twice the magic number for example rather than doing a cse while modified is true?

**Review by jfecher — CHANGES_REQUESTED** [review]

File: crates/noirc_evaluator/src/ssa/conditional.rs
Comment (guipublic) [reply crates/noirc_evaluator/src/ssa/conditional.rs:379]: Synchronisation of functions may replace function calls in one branch from a call in the other branch. The first cse update the result instructions with the reference to the replaced call, which now leaves identical result instructions (from the two branches). The second cse can now optimise them.

File: crates/noirc_evaluator/src/ssa/context.rs
Comment (guipublic) [reply crates/noirc_evaluator/src/ssa/context.rs:705]: This one is not required, but investigating changing results with the Poseidon PR, I noticed that it can be help a lot. It is expected that a cse round generates opportunities for more optimisations, so ideally we would do it until there is no gain, but I am a bit worried on the performance impact and would prefer to leave this to an optimisation setting.

File: crates/noirc_evaluator/src/ssa/optimizations.rs
Comment (guipublic) [reply crates/noirc_evaluator/src/ssa/optimizations.rs]: Done

**Review by jfecher — COMMENTED** [review]
Looks good. Lets leave some comments in the code explaining why we're calling the same functions twice in a row and then this PR should be ready

**Review by jfecher — APPROVED** [review]

**Comment by kevaundray** [comment]
Been trying to merge this, but PRs keep getting merged infront of it -- This will be fixed once we enable Merge Queues!

**Pull request merged**
